pseudocode-defns/pseudocode_defns.tex

```diff
@@ -0,0 +1,671 @@
+
+For example, \grace{TODO}.
+\end{definition}
+
+\begin{definition}[\texttt{u32}]$\sfi$
+    \texttt{u32} is the Rust 32-bit unsigned integer type. If $v$ is a value of type \texttt{u32}, then we know that $v\in \{0,1,2,\ldots, 2^{32}-1\}$. \inRust{u32}{https://doc.rust-lang.org/std/primitive.u32.html}
+\end{definition}
+
+\begin{definition}[\texttt{usize}]
+    \texttt{usize} is defined differently on 32-bit and 64-bit machines. This is because the size of this primitive is equal to the number of bytes it takes to reference any location in memory.
+    \begin{itemize}
+        \item \textbf{32-bit machines:} if \texttt{v} is a value of type \texttt{usize}, then $\texttt{v}\in \{0,1,2,\ldots,2^{32} - 1\}$
+        \item \textbf{64-bit machines:} if \texttt{v} is a value of type \texttt{usize}, then $\texttt{v}\in \{0,1,2,\ldots,2^{64} - 1\}$
+    \end{itemize}
+    \inRust{usize}{https://doc.rust-lang.org/std/primitive.usize.html}
+\end{definition}
+
+\begin{definition}[\texttt{Vec[T]}]$\sfi$
+    The Rust type \texttt{Vec[T]} consists of ordered lists of type \texttt{T}. For example, if \texttt{T = bool}, then values of type \texttt{Vec[T]} include $\texttt{[], [0], [1], [0, 0],} \ldots$. \inRust{Vec<T>}{https://doc.rust-lang.org/std/vec/struct.Vec.html}
+\end{definition}
+
+\horizline
+
+\subsection{Notes, todos, questions}
+
+%\silvia{Discuss: bounded vs lower bound and upper bound}
+
+\todonei{Include info on MPFR, and possibly relate it to our existing definitions of floats.}
+
+\todonei{Define plus, minus, etc. below each type on which they operate. For example, the definition for \texttt{u32} should also include a definition of plus on \texttt{u32}, multiplication on \texttt{u32}, etc.}
+
+\questionr{Should we have a general definition for ``floats'' (and ``integers''?), or is it sufficiently understood what a float is in general?}
+
+\section{Domains}
+
+$\sfi$ A \emph{data domain} is a representation of the set of values on which a metric or function can operate. For example, if a function accepts inputs from the domain \texttt{IntervalDomain(1:u32,17:u32)}, this means that the function can take any input value \texttt{v} of type \texttt{u32} such that \texttt{1 <= v} and \texttt{v <= 17}.
+
+\begin{definition}[\texttt{AllDomain}]$\sfi$
+\texttt{AllDomain(T)} is the domain of all values of type $\T$. \domainType{AllDomain[T]}
+\end{definition}
+
+For example, \texttt{AllDomain(u32)} is the domain of all values of type \texttt{u32}.
+
+
+\begin{definition}[\texttt{ImputableDomain}] 
+Any domain for which the \texttt{ImputableDomain} trait is implemented for it, has 
+\begin{itemize}
+    \item an associated type \texttt{NonNull}. \texttt{VectorDomain::NonNull} is the datum type after imputation. 
+    \item an imputation function \texttt{impute\_constant} that replaces a null value with a constant or passes a non-null value through
```

Comment:
the type signature and definition of impute_constant and the other associated functions like is_null and null

Reply:
Sorry, could you clarify what you meant here?